### DISCUSSION TITLE
Offline mutations order & dependencies

### DISCUSSION THREAD

**Discussion by Jylth** [urql-graphql/urql#1773] [Q&A] [answered]
Hello,

First, I would like to thank all the people working on this sweet tool, you're all doing a great job that is hugely appreciated !

I do have a question about the offline exchange and how it handles optimistic mutations: Is the order of the calls of each mutation respected, and is there a way for mutation to have dependencies on other mutations?

The dependency point seems to be primordial to handle the scenarios where a create, update & delete are all happening on one entity or when we would want to create relations of an offline entity. For example, creating an item of a parent project: the backend will probably have a foreign key constraint where the item will require its parent to have been created beforehand.

For this to work, a concept of linked mutations where each link waits for the response of the previous one before proceeding would be required. Those links would need to have access to the updated cache (or dependencies responses) at each step, so that, for example, a child item can use its parent id to be able to be created.

It seems that right now the offline exchange doesn't handle those use cases, because it will send all pending mutations at once, and so the exchange could only be used in scenarios with fire-and-forget mutations without any dependencies.

Would love to hear some opinions/advice on how to solve the use cases described above with the current state of the offline-exchange. Probably through a custom exchange filtering and managing mutations ?

**Comment by kitten** [comment, marked as answer]
that's all correct so far 👍

The order of results is always respected since it isn't safe to assume that mutations can be ordered as they were sent, since this isn't the same as with queries. Because the network and the API itself Kay cause mutations to run in any given order, we don't reorder mutations. Optimistic mutations may still stack but not have a special order

While we do take care in how queries and optimistic mutations are applied and reorder those, this hence doesn't apply to mutation results. For that were assuming that an exchange would sometimes have to be written to serialise mutation execution. Currently we're not shipping a utility or package for that though

**Reply by Jylth** [reply]
Thanks for the quick reply !

Well very good to know, I am currently working on a React Native app requiring a very strong offline behavior, so I will be working on building those missing utilities to Urql in the next weeks.

Will keep you'all up to date if I have anything worth sharing 😄

**Reply by alexjharrison** [reply]
@Jylth 

We were coming up against the same problem so I wrote an exchange that would queue all operations when the user is online (since graphcache handles the offline failedQueue list).

It sets a flag so while one operation is waiting for a response, all new ones go into a backlog array.  As results come back it fires off the next one in the list.

I put the exchange directly after offlineExchange so it can catch the stored operations when they come back online.



```javascript
import { Exchange, Operation } from "@urql/vue";
import { useOnline } from "@vueuse/core";
import { filter, onPush, pipe } from "wonka";
import { useLogger } from "../logger";

const { info } = useLogger("MUTATION_LIMITER_EXCHANGE");

const backlog: Operation[] = [];
const online = useOnline();
let inFlight = false;

export const mutationLimiterExchange: Exchange = ({
  client,
  forward
}) => operations$ => {
  const operationResult$ = pipe(
    operations$,
    filter(op => {
      
      // pass through queries, subscriptions & teardowns
      // pass through mutations if offline
      if (op.kind !== "mutation" || !online.value) return true;
      
      // if any response is pending throw on backlog
      else if (inFlight) {
        backlog.push(op);

        info("Operation discarded, Sent to backlog", { op, backlog });
        return false;
      } else {
        inFlight = true;

        info("Operation Starting, blocking new operations", { op, backlog });
        return true;
      }
    }),
    forward
  );

  return pipe(
    operationResult$,
    onPush(op => {
      // ignore non-mutation operations or offline mutation
      if (op.operation.kind !== "mutation" || !online.value) return;

      // set current operation as completed
      inFlight = false;

      if (backlog[0]) {
        // fire off next operation
        const nextOp = backlog.shift();

        if (nextOp) client.reexecuteOperation(nextOp);

        info("Operation succeeded, Firing next operation", {
          finishingOperation: op,
          newOperation: nextOp,
          backlog,
          inFlight
        });
      } else {
        info("Operation succeeded, Backlog Empty, Waiting for new requests", {
          finishingOperation: op,
          backlog,
          inFlight
        });
      }
    })
  );
};
```

**Reply by Jylth** [reply]
Good to see I'm not the only one working on this kind of issue :smile: 

I found an even simpler working solution, that requires to have a custom fetchExchange. The current fetchExchange is using the[ mergeMap operation](https://wonka.kitten.sh/api/operators#mergemap). You just need to replace that with the [concatMap operation](https://wonka.kitten.sh/api/operators#concatmap), which will actually wait for each operation to finish before starting the next one.

If you are using the multipartFetch exchange, then you should also change its behavior so that the `extract-files` function can be customized to take in a `isExtractableFile` function so that it can handle serialized `File`, `Blob` or `ReactNativeFile` instances. The current implementation only checks if an object is an instance of the classes mentioned above. You need to be able to handle serialized data of those instances, because they will be serialized by the offlineExchange.

Also, I had to make some small changes to the offlineExchange on my part for everything to work correctly when the user got back online. The `isOfflineError` check was also not perfect for a react-native environment. Will post my changes in the next days after some validation & QA by our testers.

**Reply by kitten** [reply]
@Jylth All I'd note on that is that it'll also queue queries, but in general that's also a quick fix 👍

**Reply by Jylth** [reply]
Indeed, so filtering the operations by mutation type before using the concatMap is required.
